adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,280 @@
+= Extension Structure Guidelines
+
+* Status: Proposed
+* Date: 2025-07-25
+* Authors: @cescoffier, @gsmet
+
+== Context
+
+The Quarkus ecosystem includes more than 700 extensions, many of which were developed without a consistent structure.
+The Quarkus core repository alone contains over 150 extensions.
+
+Most extensions follow the basic runtime/deployment split, but the boundaries between API and internal logic have often blurred.
+Additionally, without clear package structure rules, many split packages have emerged.
+This situation introduces several issues:
+
+* It is unclear which classes are intended as public APIs versus internal implementations.
+* There is no clear boundary between what an extension consumer can depend on and what they should not use.
+* Some extensions mix SPI and implementation code, reducing modularity and increasing the chance of unwanted coupling.
+* The need for build item re-use or coordination between extensions leads to ad-hoc patterns.
+* The word _deployment_ is there for historic reasons, but does not make a lot of sense. `build` would be better.
+
+More recently, we introduced dev-only modules, further increasing complexity.
+
+To address these issues and prepare for long-term maintainability (and possibly Java Platform Module System (JPMS) adoption), we propose a more structured layout for extension modules.
+
+== Decision
+
+Extensions may adopt a standardized structure, composed of one mandatory and several optional modules, depending on their complexity and integration needs:
+
+* `runtime`: Contains the runtime implementation.
+It may expose public APIs if needed but should prefer delegation to the `runtime-api` module (See <<the-place-of-the-public-api>>).
+This module also contains the `recorders` and other runtime logic.
+* `runtime-api`: Contains public APIs intended to be consumed by other extensions or application code.
+This module should have minimal dependencies.
+Depending on an `runtime-api` module does not transitively include the full extension (i.e., it avoids pulling in runtime or build logic). See <<coupling-api-and-spi>> for more details.
+* `build`: Contains build-time logic, including `BuildSteps`.
+This module may define internal build items.
+* `build-spi`: Contains build-time APIs intended to be reused across multiple extensions for integration purposes.
+Extensions contributing to or using build items from other extensions should depend on `build-spi` modules.
+* `dev`: Contains runtime classes used exclusively in dev mode (e.g., for the Dev UI). This avoids shipping dev-only classes into production artifacts.
```

Comment:
I forgot, did we consider calling it `runtime-dev`? Theoretically, we may want to introduce `build-dev` too.

Reply:
it makes sense to have it named `runtime-dev` as we have `runtime-api` and not just `api`.